### PR TITLE
[eth] Add parsePriceFeedUpdates method

### PR DESCRIPTION
**Pull request by ali-bahjati**
Adds the implementation and tests and benchmarks for it (on Foundry). The tests that take an input are fuzzy and run a couple of times (256 by default).

There is a corner case when updateData contains multiple updates for the same ID. The code will chose any (the first one) within the given time range and we do not guarantee anything else. It was not clear to me whether earliest is better or the oldest. 

This change is divided in two commits, first the implementation with duplicate code. The second commit tries to reduce the redundancy a little bit without sacrificing much gas. The difference is which is acceptable:

```
testBenchmarkParsePriceFeedUpdatesForTwoPriceFeed() (gas: 67 (0.021%)) 
testBenchmarkParsePriceFeedUpdatesForOnePriceFeedNotWithinRange() (gas: 73 (0.023%)) 
testBenchmarkParsePriceFeedUpdatesForOnePriceFeed() (gas: 73 (0.024%)) 
testBenchmarkUpdatePriceFeedsIfNecessaryFresh() (gas: 269 (0.071%)) 
testBenchmarkUpdatePriceFeedsFresh() (gas: 269 (0.076%)) 
testBenchmarkUpdatePriceFeedsNotFresh() (gas: 269 (0.086%)) 
```

### CODE REVIEW

File: ethereum/forge-test/Pyth.t.sol [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+    IPyth public pyth;
+
+    // -1 is equal to 0x111111 which is the biggest uint if converted back
+    uint64 constant MAX_UINT64 = uint64(int64(-1));
+    
+    function setUp() public {
+        pyth = IPyth(setUpPyth(setUpWormhole(1)));
+    }
+
+    function generateRandomPriceAttestations(
```
Comment (ali-bahjati) [inline ethereum/forge-test/Pyth.t.sol:24]: This method and the method below could be moved to the PythTestUtils but I won't move them now. I will move them if another test wants to use them (which will probably mean dividing Pyth tests into multiple files).

**Review by jayantk — APPROVED** [review]
I'm going to approve, but I don't love the code duplication for deserializing the price feed data. It sounds like you've looked into some options for eliminating it, but would be good to think about it again. The deserialization code is tricky and there's a distinct chance that we'll forget to update one copy of it if the message format ever changes.

File: ethereum/contracts/pyth/Pyth.sol [hunk 2]
```diff
@@ -267,14 +212,221 @@ abstract contract Pyth is PythGetters, PythSetters, AbstractPyth {
...
+                    encoded = vm.payload;
+                }
+
+                (uint index, uint nAttestations, uint attestationSize) = 
+                    parseBatchAttestationHeader(encoded);
+
+                bytes32 priceId;
+
+                // Deserialize each attestation
+                for (uint j=0; j < nAttestations; j++) {
```
[the remark this answers is not in the document]
Comment (ali-bahjati) [reply ethereum/contracts/pyth/Pyth.sol:322]: The problem is that one is creating PriceFeed and one is create PriceInfo. So if I want to abstract it I should convert one to each other which will result in more gas usage. 

I tried it once (getting PriceInfo and converting it to PriceFeed and it increased gas a lot. Let me try it again and give you the exact number)

**Review by njkumr — APPROVED** [review]

File: ethereum/contracts/pyth/Pyth.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (ali-bahjati) [reply ethereum/contracts/pyth/Pyth.sol:322]: I changed it (and tried to optimize it) and it is the result:
```
testBenchmarkParsePriceFeedUpdatesForOnePriceFeedNotWithinRange() (gas: 265 (0.085%)) 
testBenchmarkParsePriceFeedUpdatesForOnePriceFeed() (gas: 266 (0.086%)) 
testBenchmarkParsePriceFeedUpdatesForTwoPriceFeed() (gas: 540 (0.170%)) 
testBenchmarkUpdatePriceFeedsIfNecessaryFresh() (gas: 1125 (0.298%)) 
testBenchmarkUpdatePriceFeedsFresh() (gas: 1125 (0.316%)) 
testBenchmarkUpdatePriceFeedsNotFresh() (gas: 1373 (0.437%)) 
```

Strangely it's costing more for the updatePriceFeeds although it's the exact same logic moved to another method.

[the remark this answers is not in the document]
Comment (ali-bahjati) [reply ethereum/contracts/pyth/Pyth.sol:329]: As it's not very high I'll add it.

**Review by jayantk — APPROVED** [review]
